### PR TITLE
[asset-resources 6/n] with_resources method

### PR DESCRIPTION
**Pull request by dpeng817**
This PR allows resources to be added directly to asset defs, graph defs, jobs, schedules, and sensors. It does this using a ResourceAddable interface which communicates how to combine resource definitions.

Resource overrides are respected on each definition.

### CODE REVIEW

File: python_modules/dagster/dagster/core/asset_defs/assets.py [hunk 1]
```diff
@@ -310,6 +316,37 @@ def to_source_assets(self) -> Sequence[SourceAsset]:
...
+        ensure_requirements_satisfied(
+            merged_resource_defs,
+            list(
+                [
+                    requirement
+                    for requirement in self.get_resource_requirements()
+                    if requirement.key != "io_manager"
+                ]
+            ),
+        )
```
Comment (dpeng817) [inline python_modules/dagster/dagster/core/asset_defs/assets.py:336]: [1]

File: python_modules/dagster/dagster_tests/core_tests/resource_tests/test_with_resources.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    @asset
+    def my_derived_asset(my_source_asset):
+        return my_source_asset + 4
+
+    transformed_source, transformed_derived = with_resources(
+        [my_source_asset, my_derived_asset], resource_defs={"io_manager": the_manager}
+    )
+
+    # When an io manager definition is provided using the generic key, that generic key is used as the io manager key for the source asset.
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/core_tests/resource_tests/test_with_resources.py:113]: Nitpick: line length limit.

File: python_modules/dagster/dagster/core/asset_defs/assets.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/assets.py:336]: No need for `list` here, right?

**Review by OwenKephart — APPROVED** [review]
nice

**Pull request merged**